### ISSUE TITLE
Document the extent to which SemVer-compatibility affects deps.rs's determination of a dependency's status

### ISSUE THREAD

**Issue by 8573** [deps-rs/deps.rs#82]
As far as I see, it is currently undocumented to what extent, if any, deps.rs considers SemVer-compatibility when making determinations of whether dependencies are outdated.  I would like to suggest that this be documented.  In particular, I would appreciate the following questions' being answered in the documentation, in which _C_ is a crate and _D_ is a crate on which _C_ depends:
- If there is a version _D₂_ of _D_ that is newer than the version _D₁_ of _D_ on which _C_ depends, but _D₂_ is SemVer-incompatible with _D₁_, and _D_ has no version that is both newer than _D₁_ and SemVer-compatible with _D₁_, will deps.rs call _C_'s dependency on _D_ outdated?
  - In other words, is deps.rs okay with crates not upgrading to the very latest versions of their dependencies if upgrading would entail adapting to API changes?
- If _C_ depends on _D_ with a SemVer constraint _V_, and there is at least one newer version of _D_ than is required by _V_, but all newer versions of _D_ match _V_ (and thus are allowed to be selected by Cargo to build _C_), will deps.rs call _C_'s dependency on _D_ outdated?
  - In other words, is deps.rs okay with crates' being flexible in their dependency version specifications, as long as the latest versions of dependencies are accepted?

**Comment by Enet4** [comment]
(Please let me know if I misunderstood any of your statements, these things can get a bit tricky to expose in prose.)

To the best of my knowledge, the outdated dependency check relies on version requirements as written in the crate manifest, rather than specific versions that could have been fetched and compiled under that specification. So, it will test those requirements against the latest version of each direct dependency, in accordance to the SemVer specification. Cargo.lock is never even checked either (see #26).

> - If there is a version D₂ of D that is newer than the version D₁ of D on which C depends, but D₂ is SemVer-incompatible with D₁, and D has no version that is both newer than D₁ and SemVer-compatible with D₁, will deps.rs call C's dependency on D outdated?
>    - In other words, is deps.rs okay with crates not upgrading to the very latest versions of their dependencies if upgrading would entail adapting to API changes?

So for example, D₁ is 0.1.0 and D₂ is 0.2.0, with no other version between 0.1.0 and 0.2.0. If crate C specifies:

```toml
[dependencies]
D = "0.1.0"
```
Then the dependency D is outdated because that requirement does not encompass 0.2.0.
SemVer incompatibility does not play a role in the algorithm other than the fact that the requirement `^0.1.0` represents _API compatibility with all `0.1.x` versions_.

> - If C depends on D with a SemVer constraint V, and there is at least one newer version of D than is required by V, but all newer versions of D match V (and thus are allowed to be selected by Cargo to build C), will deps.rs call C's dependency on D outdated?
>    - In other words, is deps.rs okay with crates' being flexible in their dependency version specifications, as long as the latest versions of dependencies are accepted?

We can imagine a crate with a deliberate decision to accept any version:

```toml
[dependencies]
D = "*"
```

I believe this one would never appear as outdated. The practical consequences, on the other hand, could...  mismatch with this  green outcome. :) Which is why this requirement is often not recommended.

With all this in mind, the only thing that ought to be documented is that it checks the version requirements of each dependency as specified in the Cargo.toml file. Whether the constraint is taking advantage of SemVer compatibility guarantees or not is mostly a consequence of how SemVer was designed.

**Comment by 8573** [comment]
Okay, thanks for answering my questions.

> > * If C depends on D with a SemVer constraint V, and there is at least one newer version of D than is required by V, but all newer versions of D match V (and thus are allowed to be selected by Cargo to build C), will deps.rs call C's dependency on D outdated?
> >   
> >   * In other words, is deps.rs okay with crates' being flexible in their dependency version specifications, as long as the latest versions of dependencies are accepted?
> 
> We can imagine a crate with a deliberate decision to accept any version:
> 
> ```toml
> [dependencies]
> D = "*"
> ```

I was thinking more of—
```toml
[dependencies]
D = "1.0.0"
```
—with a _D_ 1.0.1 available.

**Comment by Enet4** [comment]
The version requirement "1.0.0" is equivalent to "^1.0.0", so that one would not be outdated. Without any operators, the version is interpreted as a caret requirement. https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#specifying-dependencies-from-cratesio
